weak_intrusive_ptr

smessmer opened the pull request:
Summary: Add weak_ptr ability to intrusive_ptr.

Differential Revision: D9039980

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+  // protected destructor. We never want to destruct intrusive_ptr_target*
+  // directly.
+  virtual ~intrusive_ptr_target() {
+    AT_CHECK(
```
Make these AT_ASSERTs, because they indicate internal library error. (There is an important difference: stack traces for AT_CHECK are suppressed, but AT_ASSERT is not). If you still want the message use AT_ASSERTM

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+
+  constexpr intrusive_ptr_target() noexcept : refcount_(0), weakcount_(0) {}
+
+ private:
```
nit: no indent here?

smessmer replied:
This file is 'arc lint' formatted

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+   * destructed by the scope (i.e. without intrusive_ptr), this function will
+   * not be called.
+   */
+  virtual void free() const {}
```
free seems like a worse method name than `release_resources`. The name `free` might suggest that you need to `delete this` in the implementation, but that's not true.

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+      std::is_base_of<intrusive_ptr_target, TTarget>::value,
+      "intrusive_ptr can only be used for classes that inherit from intrusive_ptr_target.");
+  static_assert(
+      NullType::singleton() == NullType::singleton(),
```
lol, I didn't know you could test for constexpr this way

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+      size_t new_refcount = ++target_->refcount_;
+      AT_CHECK(
+          new_refcount !=
+          1) // intrusive_ptr: Cannot increase refcount after it reached zero.
```
AT_ASSERT

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+  }
+
+  template <class From, class FromNullType>
+  intrusive_ptr(intrusive_ptr<From, FromNullType>&& rhs) noexcept
```
What do we need the convertible constructor for?

smessmer replied:
converting `intrusive_ptr<ChildClass>` to `intrusive_ptr<BaseClass>`

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+  }
+
+  size_t use_count() const noexcept {
+    if (target_ == NullType::singleton()) {
```
I guess you didn't turn on refcounting for NullType singleton in the end?

smessmer replied:
yes, that collided with the goal of allowing to use `intrusive_ptr<T>` (i.e. without a special `NullType`), which then would just use `nullptr`.

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+
+  template <class... Args>
+  static intrusive_ptr make(Args&&... args) {
+    auto result = intrusive_ptr(new TTarget(std::forward<Args>(args)...));
```
This is a bit weird to me. Why doesn't the constructor for `TTarget` start off with `refcount_ == 1 && weakcount_ == 1`?

smessmer replied:
Because I want to allow objects inheriting from `intrusive_ptr_target` to be instantiated on the stack if you don't need an `intrusive_ptr` to them, and then the refcount would semantically be zero. This is for example needed for the destructor assertion checking that refcount is zero on destruction.

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+    if (target_ != NullType::singleton()) {
+      size_t new_weakcount = ++target_->weakcount_;
+      AT_CHECK(
+          new_weakcount != 1,
```
AT_ASSERT.

Also, this one seems a bit goofy, because if weakcount really was zero before, then probably this memory had already been freed, and you're already in memory corruption land.

smessmer replied:
True. It isn't foolproof because the memory might be overwritten. However, if this assert fires, we always know something's wrong.

ezyang commented on aten/src/ATen/core/intrusive_ptr.h:
```diff
@@ -0,0 +1,532 @@
+  }
+
+  bool expired() const noexcept {
+    return use_count() == 0;
```
This will report that the NullType::singleton() is expired. I... don't think this is what you want?

smessmer replied:
isn't it? I think that's what semantically makes sense. Does it hurt one of our use cases?

smessmer committed:
IntrusivePtr

Summary: Add an IntrusivePtr class to do intrusive refcounting with a shared_ptr-like interface.

Differential Revision: D9018619

fbshipit-source-id: 83d6eae6f39dc67d98b2b45dc18e9ccdbbcb2e5e

smessmer committed:
weak_intrusive_ptr (#10038)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10038

Add weak_ptr ability to intrusive_ptr.

Differential Revision: D9039980

fbshipit-source-id: 58bd73f5704643b8bca799de1812d7aa5d4a44cb
